pkg/kubelet/token/token_manager.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package token
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+func NewManager(c corev1.CoreV1Interface) *Manager {
+	return &Manager{
+		c:     c,
+		cache: make(map[string]*authenticationv1.TokenRequest),
+		now:   time.Now,
+	}
+}
+
+type Manager struct {
+	c corev1.CoreV1Interface
+
+	// cmu guards the cache
+	cmu   sync.RWMutex
+	cache map[string]*authenticationv1.TokenRequest
+
+	// mocked for testing
+	now func() time.Time
+}
+
+func (m *Manager) GetServiceAccountToken(name, namespace string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+	key := keyFunc(name, namespace, tr)
+	ctr, ok := m.get(key)
+
+	var err error
+	if !ok || m.requiresRefresh(ctr) {
+		tr, err := m.c.ServiceAccounts(namespace).CreateToken(name, tr)
+		if err != nil {
+			glog.Errorf("couldn't update token %s: %v", key, err)
```

Comment:
utilruntime.HandleError when you eat an error

Reply:
This isn't totally eaten. If we get an error here, we may return it later if the token is expired. In the case where we are proactively rotating a token, the proactive rotation fails, but we still have a valid token, we shouldn't crash (which HandleError does in certain circumstances). In that case we want to eat the error, and return the old but still valid token.